### ISSUE TITLE
Update parser for Rust 1.0.0-alpha syntax changes

### ISSUE THREAD

**Issue by asheb** [Vektah/idea-rust#29]
Currently these snippets are not supported:

    // associated types
    impl Mul<Coord> for DeltaTime {
        type Output = Coord;
        fn mul(self, x: Coord) -> Coord { x * self.float_seconds() }
    }
    
    // where clause & if let (+ while let)
    pub fn unite<I>(mut boxes: I) -> BoundingBox where I: Iterator<Item=BoundingBox> {
        if let Simple(Text{ref mut text, ..}) = state.content {
            *text = self.update_and_get_value_text(dt);
        };
        Unknown
    }

    // where clause & closure sugar (+ with closure return type)
    pub fn with_loaded<F>(shape: &Shape, cache: &mut ShapeCache, f: F)
            where F: Fn(u32, u32, i32) {
    }

    // semicolon after macro_rules
    macro_rules! check_gl_err(
        () => (
            {
                let err = gl::GetError();
                assert!(err == gl::NO_ERROR, format!("{} or 0x{:x}", err, err));
            }
        )
    );

    // const raw pointer
    fn c_str_ptr(bytes: &[u8]) -> *const i8 {
        CString::from_slice(bytes).as_ptr()
    }

    // new fixed-size array syntax
    pub struct Matrix([MatrixColumn; 4]);
    type MatrixColumn = [Coord; 4];

    // more closure sugar
    pub type UpdateFn = FnMut(&mut FigureState, DeltaTime, InputState) + Send;

    // lifetime bounds
    pub struct Figure<'a> {
        pub state: FigureState<'a>,
        pub behavior: Option<Box<Behavior + 'a>>
    }

    // kind bounds
    pub struct Renderer {
        backend: Box<Backend + Send>,
        matrix: Matrix
    }

**Comment by asheb** [comment]
I'm starting this one. Coming back to #23 after that.

**Comment by dsyang** [comment]
@asheb! It's awesome that you're doing this! I'm curious, are you using RustSourceTest.java to find problems? I made some PRs for fixing syntax before and that's what I used.

**Comment by asheb** [comment]
Hi, @dsyang!
No, currently I'm using some my own project (plus ParserTest files to detect regressions).
That would surely nice to check parser against Rust sources, but I have some troubles with it... need to look into it...

**Comment by dsyang** [comment]
Ah ok. Well not sure how much you've looked into it, but this is how i got
them to work:
There is a method, RustSourceTest#getTestDataPath(), that needs to return
the absolute path to the src/ directory of the rust repo.  You can either
hard code it in or add a definition in the gradle.build file that points to
that src/ directory. Both options should get the Rust source tests to run.
Hope this helps! And good luck updating the parser!
On Jan 17, 2015 12:42 AM, "Alexander Sheboltaev" <notifications@github.com>
wrote:

Hi, @dsyang <https://github.com/dsyang>!
No, currently I'm using some my own project (plus ParserTest files to
detect regressions).
That would surely nice to check parser against Rust sources, but I have
some troubles with it... need to look into it...

—
Reply to this email directly or view it on GitHub
<https://github.com/Vektah/idea-rust/issues/29#issuecomment-70359576>.

**Comment by asheb** [comment]
Well, I'd set it in **gradle.properties**, but they're all skipped for some reason.

BTW, do you know how to run them from IDEA? Gradle seems to be too slow...
When I run them with JUnit Run Configuration, it complains about

    Invalid home path 'C:/Users/asheboltaev/.IntelliJIdea14/system/plugins-sandbox/test'

I've checked it: there is no **test** folder in **plugins-sandbox**. 
And later:

    java.lang.NoClassDefFoundError: Could not initialize class com.intellij.openapi.util.io.FileSystemUtil

Maybe you had something like this?

**Comment by Vektah** [comment]
`RustSourceTest` was disabled somewhere between 0.11 and 0.13. Now that Rust alpha 1.0 is out and CI is running on all new PR's we should spend some time to get it running again.

**Comment by asheb** [comment]
Well, I've removed `@Ignore` and now it complains about many folders which are no more exist in sources. Need to revisit the list of tests (actually folder names). While others are failed because of lacking support of the features I mentioned in issue description. Thinking I'll tackle this only after fixing all those features.

**Comment by Vektah** [comment]
Yeah that sounds like a good approach. 

The other thing that was in the Makefile that probably needs to be ported into the gradle test runner is a test that attempts to compile all of the `src/rust` files. This was to ensure that we don't bend the grammar to support invalid code.

On the flip side of the coin we should probably also be testing some edge cases where the grammar is known to be bad and ensuring that the grammar fails in a sane way and in roughly the right spot - The pinning in GrammarKit is great, but often the side effects can be non-obvious.

Ill create some issues for these.

**Comment by asheb** [comment]
Ok, I've added some tests, which are not compilable, while are correct syntactically (like no **main** function, or type isn't resolved). Will fix them; even so this'll add some clutter into what we actually testing.

Should we remove obsolete closure syntax entirely? Compiler gives us a nice message about that, so I'm not sure if it checks that on parsing stage:

    a.rs:2:8: 2:9 error: obsolete syntax: `|uint| -> bool` closure type syntax
    a.rs:2 fn f1(f: |i32|) -> Box<Fn(i32) + Send> {
                  ^
    note: use unboxed closures instead, no type annotation needed

**Comment by Vektah** [comment]
It looks like we can do `rustc --parse-only $file` and rustc will only generate parse errors.

Yeah I think we should get rid of old syntax for now. Post 1.0 adding conversions would be nice for bc breaks but I don't think we will need to worry about that for a while.

**Comment by dsyang** [comment]
just saw https://github.com/rust-lang/rust/pull/21452  I'm not sure what type of grammar grammarkit supports but we might be able to use the lexer and parser files to help us fix parsing bugs!

**Comment by Vektah** [comment]
@dsyang GrammarKit has some really nice features beyond standard LALR: 
 - 'top down' recursion for expression nodes
 - right recursive when you want it
 - pinning for recovery when invalid grammar is processed (not important in the rust parser, but very important in an editor)

https://github.com/JetBrains/Grammar-Kit/blob/master/HOWTO.md

Maybe we can take some inspiration from that grammar though.

**Comment by asheb** [comment]
Awesome! Inspiration definitely will be helpful :)
